Support torch.jit.ignore as a context manager

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55172 torch.jit.ignore as a context manager**

Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

tugsbayasgalan committed:
[WIP] ignore as context manager

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] ignore as context manager"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] ignore as context manager"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] ignore as context manager"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] ignore as context manager"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] ignore as context manager"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] ignore as context manager"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] ignore as context manager"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] ignore as context manager"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -291,6 +292,20 @@ def _forward(self):
     return build_def(ctx, fn_def, type_line, def_name, self_name=self_name)
 
+def is_torch_jit_ignore_context_manager(stmt):
+    # checks if the statement is torch.jit.ignore context manager
+    if isinstance(stmt.items[0].context_expr, ast.Call):
+        # extract torch part
+        function = stmt.items[0].context_expr.func
+        if isinstance(function, ast.Attribute):
+            attr_name = function.attr
+            attr_value = function.value
+            if attr_name == "ignore" and isinstance(attr_value, ast.Attribute):
+                # there should be at most two nested attributes (e.g torch.jit.ignore)
+                if attr_value.attr == "jit" and isinstance(attr_value.value, ast.Name):
+                    if attr_value.value.id == "torch":
+                        return True
+    return False
```
It is OK to fix this in next PR:

This is a bit fragile. It will match the common case of `with torch.jit.ignore`. But it would fail for following cases:

```
cm = torch.jit.ignore(...)
with cm:
    ......
```

```
import torch as th
with th.jit.ignore(...):
   ......
```

```
from torch.jit import ignore
with ignore(...):
    ......
```

I think we can try plumb a resolution callback to here in order to correctly parse `ignore`.

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -363,6 +378,94 @@ def build_param(ctx, py_arg, self_name, kwarg_only):
+        for arg in args:
+            var_name = arg.arg
+            var_ann = arg.value.value
+            var_decl_type, var_ann = var_ann.split(":")
```
We can do this in next PR:

We should add some checks here (or even earlier) to make sure the inputs and output specifications are valid

tugsbayasgalan replied:
will update in the next PR when we come up with better inp/out annotation syntax

gmagogsfm replied:
Please add a TODO before committing.

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -363,6 +378,94 @@ def build_param(ctx, py_arg, self_name, kwarg_only):
+    def create_unique_name_ext(ctx, stmt):
+        # extension will be based on the filename plus the line number
+        # of original context manager
+        return os.path.basename(ctx.filename).replace(".", "_") + "_" + str(stmt.lineno)
```
I'd rather we use the full path (after replacing all illegal chars) instead of just filename, which can still collide.

tugsbayasgalan replied:
sounds good

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -363,6 +378,94 @@ def build_param(ctx, py_arg, self_name, kwarg_only):
+        return_type_ann = ""
+        return_statement_str = "return "
+        if len(outputs) == 0:
+            return_type_ann += " -> None: pass"
```
Could you refactor the code a bit to make `pass` not part of `return_type_ann`, which is supposed to be just annotation

tugsbayasgalan replied:
yep

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -363,6 +378,94 @@ def build_param(ctx, py_arg, self_name, kwarg_only):
+        # of original context manager
+        return os.path.basename(ctx.filename).replace(".", "_") + "_" + str(stmt.lineno)
+
+    def build_return_ann_stmt(outputs):
```
Consider using NamedTuple for outputs type so that we don't need to find arg name and arg type by indices like [0] and [1]. You may have "smelled" that the use of [0] and [1] is a bit excessive in implementation, trust your smell to tell how to write perfect code : )

tugsbayasgalan replied:
hha sounds good

tugsbayasgalan replied:
done

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -363,6 +378,94 @@ def build_param(ctx, py_arg, self_name, kwarg_only):
+        for arg in args:
+            var_name, _ = arg
+            result += var_name + ", "
+        return result[:-2]
```
Why [:-2]? Consider finding a way to avoid magic numbers.

tugsbayasgalan replied:
sounds good

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -363,6 +378,94 @@ def build_param(ctx, py_arg, self_name, kwarg_only):
+
+    def build_args(args):
+        result = ""
+        for arg in args:
+            var_name, _ = arg
+            result += var_name + ", "
```
Consider using list comprehension and `join` instead?

tugsbayasgalan replied:
sounds good

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -363,6 +378,94 @@ def build_param(ctx, py_arg, self_name, kwarg_only):
+
+    # registers the custom function in the global context
+    ignore_func_str = ast.unparse(ignore_function)
+    ignore_func_str += "\nglobals()[\"{}\"] = {}".format(ignore_function_name, ignore_function_name)
+    ignore_func_str += "\nglobals()[\"{}\"]._torchscript_modifier = FunctionModifiers.IGNORE".format(ignore_function_name)
```
Why does this need to be part of string passed to `exec`? Can you run it after `exec` function definition?

tugsbayasgalan replied:
Yeah i can. I wanted to put together all the weird code in one place lol

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -363,6 +378,94 @@ def build_param(ctx, py_arg, self_name, kwarg_only):
+
+    # registers the custom function in the global context
+    ignore_func_str = ast.unparse(ignore_function)
+    ignore_func_str += "\nglobals()[\"{}\"] = {}".format(ignore_function_name, ignore_function_name)
+    ignore_func_str += "\nglobals()[\"{}\"]._torchscript_modifier = FunctionModifiers.IGNORE".format(ignore_function_name)
```
Maybe better to just generate a call to `torch.jit.ignore` decorator. That would make sure future changes we make to `torch.jit.ignore` automatically applies to functions we generate.

tugsbayasgalan replied:
Oh that is a good idea.

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -544,6 +647,12 @@ def build_Continue(ctx, stmt):
         r = ctx.make_range(stmt.lineno, stmt.col_offset, stmt.col_offset + len("with"))
+        # Handle ignore context manager
+        if is_torch_jit_ignore_context_manager(stmt):
+            if sys.version_info < (3, 9):
```
Is there a work-around or a different library to use instead of `ast.unparse` so that we can support <3.9?

tugsbayasgalan replied:
Work around is to use astunparse library, but you will have to install it to use it. Not sure if we want that?

gmagogsfm replied:
Is `astunparse` completely equivalent to `ast.unparse`?

I am weighing two sides here:

Upon launch only some users who can afford to update to 3.9+ enjoys this feature

vs

Using `astunparse` to allow all users to enjoy this feature but we risk causing some functionality differences when migrating from `astunparse` to `ast.unparse` when our python support horizon moves past 3.9

tugsbayasgalan replied:
Yeah I think they are same. I was more worried about adding new dependency into the system. Do you think it is fine?

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27895283).

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27895283).

gmagogsfm commented on test/jit/test_ignore_context_manager.py:
```diff
@@ -0,0 +1,87 @@
+        self.assertEqual(s(), model())
+
+    @unittest.skipIf(sys.version_info < (3, 9), "only supported in python3.9")
+    def test_with_ignore_context_manager_with_just_inp(self):
```
Could you add test for "just output" as well?

tugsbayasgalan replied:
done

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan commented on torch/_jit_internal.py:
```diff
@@ -540,6 +540,14 @@ def forward(self, x):
     return fn
 
+# No op context manager from python side
+class ignore_experimental(contextlib.AbstractContextManager):
```
done

tugsbayasgalan commented on torch/jit/frontend.py:
```diff
@@ -299,6 +300,21 @@ def _forward(self):
 
     return build_def(ctx, fn_def, type_line, def_name, self_name=self_name, pdt_arg_types=pdt_arg_types)
 
+def is_torch_jit_ignore_context_manager(stmt):
```
done

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27895283).

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27895283).

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27895283).

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -18,9 +19,15 @@
+from torch._jit_internal import SourceContext, should_drop, is_static_fn, FunctionModifiers  # noqa: F401
 import torch.jit.annotations
 
+_IS_ASTUNPARSE_INSTALLED = True
+try:
+    import astunparse  # type: ignore
+except ImportError:
+    _IS_ASTUNPARSE_INSTALLED = False
```
Can we make the logic a bit more defensive? ie, default value of `_IS_ASTUNPARSE_INSTALLED` is `False` and only set to true when import succeeds.

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -563,6 +665,12 @@ def build_Continue(ctx, stmt):
+        # Handle ignore context manager
+        if is_torch_jit_ignore_context_manager(stmt):
+            if not _IS_ASTUNPARSE_INSTALLED:
+                raise RuntimeError("torch.jit._IgnoreContextManager is not supported in this platform")
```
```suggestion
                raise RuntimeError("torch.jit._IgnoreContextManager requires installing Python library `astunparse`, please install it in your Python environment")
```

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27895283).

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27895283).

janeyx99 commented on tools/coverage_plugins_package/src/coverage_plugins/jit_plugin.py:
```diff
@@ -41,9 +41,13 @@ def dynamic_context(self, frame):
+                    try:
+                        sourcelines, starting_lineno = getsourcelines(obj)
+                    except OSError:
+                        pass
```
Could you add some documentation here as of why we have to bypass the error? Either with a comment or a printed warning?

tugsbayasgalan replied:
Ok I will add a comment because warning would expose our implementation details to user i think

tugsbayasgalan replied:
done

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27895283).

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27895283).

janeyx99 commented on tools/coverage_plugins_package/src/coverage_plugins/jit_plugin.py:
```diff
@@ -41,9 +41,17 @@ def dynamic_context(self, frame):
-                    sourcelines, starting_lineno = getsourcelines(obj)
-                    line_data = {filename: range(starting_lineno, starting_lineno + len(sourcelines))}
-                    cov_data.add_lines(line_data)
+                    # TODO: Because torch.jit._IgnoreContextManager relies on
+                    # python `exec` method that doesn't generate source codelines
+                    # , this part of code fails. Therefore, we just ignore the exception
+                    # until we figure out a better way to implement torch.jit._IgnoreContextManager.
```
```suggestion
                    # TODO: Because torch.jit._IgnoreContextManager relies on Python's `exec` method which doesn't 
                    # generate source codelines, getsourcelines(obj) fails. For now, we just ignore the exception
                    # until we figure out a better way to implement torch.jit._IgnoreContextManager.
```
for more clarity

tugsbayasgalan committed:
Update on "torch.jit.ignore as a context manager"


Description:
This is part 1 of series of PRs for supporting torch.jit.ignore as context manager. Following features are implemented in this PR:

- Unique name for the registered function under torch.jit.frontend module. The unique name is generated based on the file name and line number of context manager
- Forcing user to explicitly annotate the input and outputs. 
- No side effects are considered.

Differential Revision: [D27895283](https://our.internmc.facebook.com/intern/diff/D27895283)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27895283).
